### PR TITLE
DisputableVoting: Optimize quiet ending

### PR DESCRIPTION
**Pull request by facuspagnuolo**
This PR introduces an improvement to the quiet ending feature. The main purpose is to compare the result at the beginning and at the end of the quiet ending period, and extend it only if it differs.

### CODE REVIEW

File: apps/voting-disputable/test/contracts/voting_quiet_ending.js [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+            await voting.vote(voteId, currentSupport, { from: holder40 })
+          })
+
+          itStoresTheSnapshot(previousSupport)
+          itDoesNotExtendTheVoteDuration()
+        })
+      })
+    })
+  })
+})
```
Comment (bingen) [inline apps/voting-disputable/test/contracts/voting_quiet_ending.js:309]: Nice tests!!

File: apps/voting-disputable/contracts/DisputableVoting.sol [hunk 2]
```diff
@@ -871,6 +873,27 @@ contract DisputableVoting is IForwarder, DisputableAragonApp {
...
+        VoterState snapshotSupport = vote_.quietEndingSnapshotSupport;
+        if (snapshotSupport == VoterState.Absent) {
+            return false;
+        }
+
+        // Otherwise, we have to know if the vote was flipped comparing if was accepted or not during its last extension
+        bool wasAccepted = snapshotSupport == VoterState.Yea;
+        Setting storage setting = settings[vote_.settingId];
+        uint256 currentExtensions = vote_.quietEndingExtendedSeconds / setting.quietEndingExtension;
+        for (uint256 i = 0; i < currentExtensions; i++) {
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply apps/voting-disputable/contracts/DisputableVoting.sol:891]: Saving the snapshot for each extension is not trivial, so let's do the other one

File: apps/voting-disputable/contracts/DisputableVoting.sol [hunk 3]
```diff
@@ -947,58 +969,65 @@ contract DisputableVoting is IForwarder, DisputableAragonApp {
...
+        _computeCastVote(vote_, _voteId, _supports, _voter, _caster);
+
+        if (vote_.quietEndingSnapshotSupport == VoterState.Absent) {
+            // If the snapshot support at beginning of the quiet ending period was not taken yet, we simply store it. Note that if there
+            // were no votes during the quiet ending period, this snapshot is never stored and the vote end date is not affected.
+            vote_.quietEndingSnapshotSupport = _voterStateFor(wasAccepted);
+        } else {
+            // First, we make sure the extension is persisted, if are voting within the extension and it was not considered yet, we store it.
+            // Note that we are trusting `_canVote`, if we reached this point, it means the vote's flip was already confirmed.
+            if (getTimestamp64() >= _finalVoteEndDate(vote_)) {
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply apps/voting-disputable/contracts/DisputableVoting.sol:1026]: I see, honestly I couldn't come up with a clearer but similarly optimized as this one to compute multiple extensions

File: apps/voting-disputable/contracts/DisputableVoting.sol [hunk 4]
```diff
@@ -947,58 +969,65 @@ contract DisputableVoting is IForwarder, DisputableAragonApp {
...
     /**
-    * @dev Private function to compute a cast vote
+    * @dev Private function to compute a cast vote in the normal voting phase
     *      It assumes the pointer to the vote is valid
     */
-    function _computeCastVote(Vote storage vote_, uint256 _voteId, uint256 _yeas, uint256 _nays, bool _supports, address _voter, address _caster)
-        private
-        returns (uint256, uint256)
-    {
+    function _computeCastVote(Vote storage vote_, uint256 _voteId, bool _supports, address _voter, address _caster) private {
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply apps/voting-disputable/contracts/DisputableVoting.sol:982]: Yea, I've been thinking about this as well, I think we simply carried it from the traditional Voting app, but I agree to use the prefix as we are doing for other variables

**Comment by facuspagnuolo** [comment]
@bingen thx a lot for the review! let me know your thoughts :D

File: apps/voting-disputable/contracts/DisputableVoting.sol [hunk 5]
```diff
@@ -256,10 +256,10 @@ contract DisputableVoting is IForwarder, DisputableAragonApp {
     * @param _supports Whether voter supports the vote
     */
     function vote(uint256 _voteId, bool _supports) external {
-        Vote storage vote_ = _getVote(_voteId);
-        require(_canVote(vote_, msg.sender), ERROR_CANNOT_VOTE);
+        Vote storage _vote = _getVote(_voteId);
```
Comment (bingen) [inline apps/voting-disputable/contracts/DisputableVoting.sol:259]: I meant only for function params, but this looks fine to me too.

**Pull request merged**